### PR TITLE
 Using user-defined Javascript to customize geospatial visualization

### PR DESCRIPTION
**Pull request by mistercrunch**
screen shot 2018-01-08 at 10 09 46 am

Proof of concept on custom user-provided JS on `deck_path` an `deck_scatter` visualizations. Users can pick extra data columns that become available in their javascript functions' scope.

Users can pass 3 javascript snippet to respectively:
* intercept the object corresponding to the data point and mutate it. This enables them to do things like controlling the stroke-color, stroke-width, fill-color, fill-opacity either in a fixed way or to bind it to related data
* write a JS function that returns an HTML tooltip, allowing user to craft rich tooltips that expose many attributes of the data
* write a JS function that returns a url that defines the onclick event

### CODE REVIEW

File: superset/assets/visualizations/deckgl/layers/geojson.jsx [hunk 1]
```diff
@@ -23,11 +25,11 @@ const convertGeoJsonColorProps = (p, colors) => {
...
 };
 
-export default function geoJsonLayer(formData, payload) {
+export default function geoJsonLayer(formData, payload, slice) {
   const fd = formData;
   const fc = fd.fill_color_picker;
   const sc = fd.stroke_color_picker;
-  const data = payload.data.geojson.features.map(d => ({
+  let data = payload.data.geojson.features.map(d => ({
     ...d,
```
Comment (hughhhh) [inline superset/assets/visualizations/deckgl/layers/geojson.jsx]: Why let instead of const?

Comment (betodealmeida) [reply superset/assets/visualizations/deckgl/layers/geojson.jsx]: Because now `data` can be mutated (see below, line 44).

File: superset/viz.py [hunk 2]
```diff
@@ -1876,8 +1882,12 @@ def get_data(self, df):
 
         features = []
         for d in df.to_dict(orient='records'):
-            d = dict(position=self.get_position(d), **self.get_properties(d))
-            features.append(d)
+            print(d)
```
Comment (hughhhh) [inline superset/viz.py]: remove print

File: superset/viz.py [hunk 3]
```diff
@@ -1949,22 +1959,22 @@ class DeckPathViz(BaseDeckGLViz):
...
-        paths = [deser(s) for s in df[fd.get('line_column')]]
+        path = deser(d[fd.get('line_column')])
         if fd.get('reverse_long_lat'):
-            paths = [[(point[1], point[0]) for point in path] for path in paths]
-        d = {
-            'mapboxApiKey': config.get('MAPBOX_API_KEY'),
-            'paths': paths,
+            path = (path[1], path[0])
+        return {
+            'path': path,
```
Comment (hughhhh) [inline superset/viz.py]: nit: 
set the return to just have `'path': (path[1], path[0])`
```
return {
             'path': (path[1], path[0]),
}
```

**Comment by hughhhh** [comment]
🚢

**Review by betodealmeida — APPROVED** [review]
<3 this

File: superset/viz.py (hunk 2 above)
Comment (mistercrunch) [reply superset/viz.py]: oops

File: superset/viz.py [hunk 4]
```diff
@@ -1949,22 +1959,22 @@ class DeckPathViz(BaseDeckGLViz):
 
     def query_obj(self):
         d = super(DeckPathViz, self).query_obj()
-        d['groupby'] = []
-        d['metrics'] = []
-        d['columns'] = [self.form_data.get('line_column')]
+        line_col = self.form_data.get('line_column')
+        if d['metrics']:
+            d['groupby'] += [line_col]
```
[the remark this answers is not in the document]
Comment (hughhhh) [reply superset/viz.py]: shorthand all day 💯  @betodealmeida

**Pull request merged**